src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -295,6 +295,8 @@ def _bid(self, payload):
 
             arguments['buyer'] = self.device.name
             arguments['buyer_origin'] = self.device.name
+            arguments['buyer_origin_id'] = self.device.uuid
+            assert self.can_bid_be_posted(self.next_market.time_slot, **arguments)
```

Comment:
I agree that is should be here. But please remove it from `_bid_impl `

Reply:
Same goes for `can_offer_be_posted` actually.

Reply:
Or shall we check only shortly before the bid is added?
In any case, this needs to be homogenised.

Reply:
I would prefer to be in `_bid/_offer` than `_bid_impl/_offer_impl` as `can_bid_be_posted /can_offer_be_posted ` inside `_bid/_offer` shows the external user that their bids/offer can no more be posted for that slot

Reply:
I agree with you, however, if you look into the load and pv, these checks are done in `bid_impl` `offer_impl`.
There mus be a reason for that, but i can not remember. @spyrostz Can you help us please?

Reply:
In order to avoid race conditions with other bids / offers. _bid/_offer are executed in the context of the redis messaging thread (therefore not during the tick event, but between ticks). Therefore if you do the validation in these methods, there might be more bids/offers that are received before the event tick, and the energy limits of the device can be surpassed (even though the can_bid_be_posted allows us, it only takes into account bids that are already placed in the market, not bids that live in the command buffer). Thus we are obliged to do this check in the bid_impl/offer_impl, which are executed on the main thread, in the tick event. Hope it clarifies things.

Reply:
Do you think its better to have `can_bid_be_posted/can_offer_be_posted`  in both `_bid/_offer` as well as in `_bid_impl/_offer_impl` . Although the later is for sure better(after your explanation) to prevent state violation inside d3a but external client isn’t notified as the recently posted bid would have already violated the state(considering that `replace_existing=False`)?

Reply:
I think it makes thing much harder if we have them in the _bid / _offer methods, because we have to think about bid / offer deletions too. Therefore, we might reject a bid / offer that could in reality be accepted (because a bid was deleted before that, thus decreasing the total bid energy)